decoder/vaapidecoder_h265.cpp

```diff
@@ -0,0 +1,1784 @@
+{
+  /* Fixme: Add limit check based on Annex A */
+  return MAX (1, (sps->max_dec_pic_buffering_minus1[0] + 1));
+}
+
+Decode_Status VaapiDecoderH265::ensure_dpb (H265SPS * sps)
+{
+    uint32_t dpb_size;
+    dpb_size = get_max_dec_frame_buffering (sps);
+    if (m_dpb_size < dpb_size) {
+    }
+
+    /* Reset DPB */
+    if (!dpb_reset (dpb_size))
+        return DECODE_FAIL;
+
+    return DECODE_SUCCESS;
+}
+
+/* 8.3.1 - Decoding process for picture order count */
+void VaapiDecoderH265::init_picture_poc (PicturePtr& picture, H265SliceHdr *slice_hdr, H265NalUnit * nalu)
+{
+    H265PPS *const pps = slice_hdr->pps;
+    H265SPS *const sps = pps->sps;
+
+    const uint32_t MaxPicOrderCntLsb =
+    1 << (sps->log2_max_pic_order_cnt_lsb_minus4 + 4);
+    uint8_t nal_type = nalu->type;
+    uint8_t temporal_id = nalu->temporal_id_plus1 - 1;
+
+    m_prev_poc_lsb = m_poc_lsb;
+    m_prev_poc_msb = m_poc_msb;
+
+    if (!(nal_is_irap (nal_type) && picture->m_NoRaslOutputFlag)) {
+        m_prev_poc_lsb = m_prev_tid0pic_poc_lsb;
+        m_prev_poc_msb = m_prev_tid0pic_poc_msb;
+    }
+
+    /* Finding PicOrderCntMsb */
+    if (nal_is_irap (nal_type) && picture->m_NoRaslOutputFlag)
+        m_poc_msb = 0;
+    else {
+        /* (8-1) */
+        if ((slice_hdr->pic_order_cnt_lsb < m_prev_poc_lsb) &&
+           ((m_prev_poc_lsb - slice_hdr->pic_order_cnt_lsb) >=
+            (MaxPicOrderCntLsb / 2)))
+            m_poc_msb = m_prev_poc_msb + MaxPicOrderCntLsb;
+
+        else if ((slice_hdr->pic_order_cnt_lsb > m_prev_poc_lsb) &&
+            ((slice_hdr->pic_order_cnt_lsb - m_prev_poc_lsb) >
+            (MaxPicOrderCntLsb / 2)))
+            m_poc_msb = m_prev_poc_msb - MaxPicOrderCntLsb;
+        else
+            m_poc_msb = m_prev_poc_msb;
+    }
+
+    /* (8-2) */
+    m_poc = picture->m_poc = m_poc_msb + slice_hdr->pic_order_cnt_lsb;
+    m_poc_lsb = picture->m_poc_lsb = slice_hdr->pic_order_cnt_lsb;
+
+    if (nal_is_idr (nal_type)) {
+        picture->m_poc = 0;
+        picture->m_poc_lsb = 0;
+        m_poc_lsb = 0;
+        m_poc_msb = 0;
+        m_prev_poc_lsb = 0;
+        m_prev_poc_msb = 0;
+        m_prev_tid0pic_poc_lsb = 0;
+        m_prev_tid0pic_poc_msb = 0;
+    }
+
+    if (!temporal_id && !nal_is_rasl (nal_type) &&
+    !nal_is_radl (nal_type) && nal_is_ref (nal_type)) {
+        m_prev_tid0pic_poc_lsb = slice_hdr->pic_order_cnt_lsb;
+        m_prev_tid0pic_poc_msb = m_poc_msb;
+    }
+}
+
+void VaapiDecoderH265::vaapi_picture_h265_set_reference(VaapiDecPictureH265* picture, uint32_t referenceFlags)
+{
+    VAAPI_PICTURE_FLAG_UNSET(picture, VAAPI_PICTURE_FLAGS_RPS_ST | VAAPI_PICTURE_FLAGS_RPS_LT);
+    VAAPI_PICTURE_FLAG_UNSET(picture, VAAPI_PICTURE_FLAGS_REFERENCE);
+    VAAPI_PICTURE_FLAG_SET(picture, referenceFlags);
+}
+
+
+/* Get the dpb picture having the specifed poc or poc_lsb */
+VaapiDecPictureH265 *VaapiDecoderH265::dpb_get_picture (int32_t poc, bool match_lsb)
```

Comment:
Try use a nest class for all dpb.
You can refer to https://github.com/xuguangxin/libyami/blob/grid/examples/grid.cpp. Flipper is a nest class of DrmRenderer